File: compression/wavelet.py
"""Wavelet-based compression using 2D discrete wavelet transforms (PyWavelets).

Compared with the ESMI Slack `wavelet_compress.py` script (ap-atul/wavelets + db4
full-coefficient NPZ), this module keeps the research knobs we need
(keep-fraction / levels) while adopting the better reconstruction choices:

  - Default basis **db4** (same as the Slack script; far better energy
    compaction on satellite rasters than Haar)
  - True 2D DWT via ``pywt.wavedec2`` (proper for images; the Slack helper
    applies ``FastWaveletTransform`` per band)
  - Always retain the approximation (LL) sub-band, then keep the largest
    detail coefficients up to ``keep_fraction`` of all coefficients — much
    higher PSNR at the same sparsity than hard-thresholding the whole array
  - float32 transform path (matches the Slack script's memory profile)

Cloud Run / Streamlit call this module; the browser engine uses a fixed Haar
implementation with the same LL-preserving policy.
"""

from __future__ import annotations

from time import perf_counter

import numpy as np
import pywt

from compression.base import CompressionExecutionResult, build_execution_result

# Prefer the Slack-script default; Haar remains available for A/B tests.
DEFAULT_WAVELET = "db4"
SUPPORTED_WAVELETS = ("haar", "db2", "db4", "sym2", "sym4")


def _approx_mask(coeff_array: np.ndarray, coeff_slices: list) -> np.ndarray:
    """Boolean mask of the approximation (LL) coefficients in the packed array."""
    mask = np.zeros(coeff_array.shape, dtype=bool)
    approx = coeff_slices[0]
    mask[approx] = True
    return mask


def compress_band(
    band: np.ndarray,
    *,
    wavelet: str,
    level: int,
    keep_fraction: float,
) -> tuple[np.ndarray, int]:
    """
    2D wavelet decompose → keep LL + top detail coeffs → reconstruct.

    Returns the reconstructed band and the count of retained coefficients.
    """
    wavelet_name = wavelet if wavelet in SUPPORTED_WAVELETS else DEFAULT_WAVELET
    band_f = band.astype(np.float32, copy=False)
    wavelet_obj = pywt.Wavelet(wavelet_name)
    max_level = pywt.dwt_max_level(min(band.shape), wavelet_obj.dec_len)
    safe_level = max(1, min(int(level), max_level if max_level > 0 else 1))
    keep_fraction = float(np.clip(keep_fraction, 0.001, 1.0))

    coeffs = pywt.wavedec2(band_f, wavelet=wavelet_obj, level=safe_level)
    coeff_array, coeff_slices = pywt.coeffs_to_array(coeffs)
    approx_mask = _approx_mask(coeff_array, coeff_slices)
    approx_count = int(approx_mask.sum())

    # Budget: at least every LL coeff, then fill up to keep_fraction of all coeffs.
    keep_count = max(approx_count, int(np.ceil(coeff_array.size * keep_fraction)))
    keep_count = min(keep_count, coeff_array.size)

    magnitudes = np.abs(coeff_array)
    # Force LL into the retained set by ranking it above every detail coeff.
    ranking = magnitudes.copy()
    ranking[approx_mask] = np.inf
    flat = ranking.ravel()
    if keep_count >= flat.size:
        compressed_array = coeff_array.copy()
    else:
        threshold = np.partition(flat, -keep_count)[-keep_count]
        compressed_array = np.where(
            (magnitudes >= threshold) | approx_mask,
            coeff_array,
            0.0,
        )

    retained_count = int(np.count_nonzero(compressed_array))
    compressed_coeffs = pywt.array_to_coeffs(
        compressed_array, coeff_slices, output_format="wavedec2"
    )
    reconstructed = pywt.waverec2(compressed_coeffs, wavelet=wavelet_obj)
    reconstructed = reconstructed[: band.shape[0], : band.shape[1]]

    # Match original dynamic range (satellite DN / reflectance).
    lo = float(np.min(band))
    hi = float(np.max(band))
    reconstructed = np.clip(reconstructed, lo, hi)
    return reconstructed.astype(band.dtype, copy=False), retained_count


def run_wavelet_compression(
    bands: dict[str, np.ndarray],
    *,
    wavelet: str = DEFAULT_WAVELET,
    level: int = 3,
    keep_fraction: float = 0.08,
) -> CompressionExecutionResult:
    """Compress all bands independently with LL-preserving wavelet thresholding."""
    start = perf_counter()
    reconstructed_bands: dict[str, np.ndarray] = {}
    retained_total = 0
    wavelet_name = wavelet if wavelet in SUPPORTED_WAVELETS else DEFAULT_WAVELET

    for name, band in bands.items():
        reconstructed, retained_count = compress_band(
            band,
            wavelet=wavelet_name,
            level=level,
            keep_fraction=keep_fraction,
        )
        reconstructed_bands[name] = reconstructed
        retained_total += retained_count

    runtime_seconds = perf_counter() - start
    # float32-ish payload estimate (value + index overhead folded into 8 bytes).
    return build_execution_result(
        method="wavelet",
        bands=bands,
        reconstructed_bands=reconstructed_bands,
        compressed_bytes_estimate=retained_total * 8,
        runtime_seconds=runtime_seconds,
        metadata={
            "wavelet": wavelet_name,
            "level": int(level),
            "keep_fraction": float(keep_fraction),
            "preserve_approximation": True,
            "backend": "pywt.wavedec2",
        },
    )

File: src/lib/compression/wavelet.ts
import type { BandMap } from '../types'

/**
 * Browser wavelet compression — fixed multilevel Haar with LL preserved.
 *
 * The previous implementation re-transformed the *entire* image at every level
 * (incorrect). True multilevel DWT only recurses into the LL quadrant.
 * Approximation coefficients are always kept; ``keepFraction`` then budgets the
 * largest detail coefficients (same policy as the Python/db4 path).
 *
 * Daubechies-4 lives on Cloud Run / PyWavelets; Haar is the portable browser basis.
 */

function haarForward(row: Float64Array): Float64Array {
  const n = row.length
  const out = new Float64Array(n)
  const half = Math.floor(n / 2)
  for (let i = 0; i < half; i++) {
    const a = row[i * 2] ?? row[i * 2 - 1] ?? 0
    const b = row[i * 2 + 1] ?? a
    out[i] = (a + b) / 2
    out[half + i] = (a - b) / 2
  }
  // Keep the unpaired sample at the end — never overwrite the first detail at `half`.
  if (n % 2 === 1) out[n - 1] = row[n - 1]
  return out
}

function haarInverse(row: Float64Array): Float64Array {
  const n = row.length
  const out = new Float64Array(n)
  const half = Math.floor(n / 2)
  for (let i = 0; i < half; i++) {
    const approx = row[i]
    const detail = row[half + i] ?? 0
    out[i * 2] = approx + detail
    if (i * 2 + 1 < n) out[i * 2 + 1] = approx - detail
  }
  if (n % 2 === 1) out[n - 1] = row[n - 1]
  return out
}

/** In-place 2D Haar on the top-left region [0, regionH) × [0, regionW). */
function haar2DRegion(
  data: Float64Array,
  stride: number,
  regionW: number,
  regionH: number,
  inverse: boolean,
): void {
  const op = inverse ? haarInverse : haarForward

  // Rows
  for (let y = 0; y < regionH; y++) {
    const row = new Float64Array(regionW)
    const base = y * stride
    for (let x = 0; x < regionW; x++) row[x] = data[base + x]
    const transformed = op(row)
    for (let x = 0; x < regionW; x++) data[base + x] = transformed[x]
  }

  // Columns
  for (let x = 0; x < regionW; x++) {
    const col = new Float64Array(regionH)
    for (let y = 0; y < regionH; y++) col[y] = data[y * stride + x]
    const transformed = op(col)
    for (let y = 0; y < regionH; y++) data[y * stride + x] = transformed[y]
  }
}

function maxSafeLevels(width: number, height: number): number {
  let levels = 0
  let w = width
  let h = height
  while (w >= 2 && h >= 2) {
    levels += 1
    w = Math.floor(w / 2)
    h = Math.floor(h / 2)
  }
  return Math.max(1, levels)
}

function compressBandWavelet(
  band: Float64Array,
  width: number,
  height: number,
  keepFraction: number,
  levels: number,
): { reconstructed: Float64Array; retained: number } {
  const safeLevels = Math.max(1, Math.min(levels, maxSafeLevels(width, height)))
  const coeffs = new Float64Array(band)

  // Forward multilevel DWT (LL-only recursion).
  let regionW = width
  let regionH = height
  for (let level = 0; level < safeLevels; level++) {
    haar2DRegion(coeffs, width, regionW, regionH, false)
    regionW = Math.floor(regionW / 2)
    regionH = Math.floor(regionH / 2)
    if (regionW < 1 || regionH < 1) break
  }

  const llW = Math.max(1, Math.floor(width / 2 ** safeLevels))
  const llH = Math.max(1, Math.floor(height / 2 ** safeLevels))
  const approxCount = llW * llH
  const keepCount = Math.min(
    coeffs.length,
    Math.max(approxCount, Math.ceil(coeffs.length * Math.min(1, Math.max(0.001, keepFraction)))),
  )

  const magnitudes = new Float64Array(coeffs.length)
  for (let i = 0; i < coeffs.length; i++) magnitudes[i] = Math.abs(coeffs[i])

  // Rank details; force LL above every detail so it is always retained.
  const ranking = new Float64Array(magnitudes)
  for (let y = 0; y < llH; y++) {
    for (let x = 0; x < llW; x++) {
      ranking[y * width + x] = Number.POSITIVE_INFINITY
    }
  }
  const sorted = Float64Array.from(ranking).sort()
  const threshold = sorted[sorted.length - keepCount] ?? 0

  const sparse = new Float64Array(coeffs.length)
  let retained = 0
  for (let i = 0; i < coeffs.length; i++) {
    const y = Math.floor(i / width)
    const x = i % width
    const inLL = y < llH && x < llW
    if (inLL || Math.abs(coeffs[i]) >= threshold) {
      sparse[i] = coeffs[i]
      retained++
    }
  }

  // Inverse multilevel DWT.
  for (let level = safeLevels - 1; level >= 0; level--) {
    const w = Math.max(1, Math.floor(width / 2 ** level))
    const h = Math.max(1, Math.floor(height / 2 ** level))
    haar2DRegion(sparse, width, w, h, true)
  }

  let bmin = Infinity
  let bmax = -Infinity
  for (let i = 0; i < band.length; i++) {
    const v = band[i]
    if (v < bmin) bmin = v
    if (v > bmax) bmax = v
  }
  for (let i = 0; i < sparse.length; i++) {
    sparse[i] = Math.min(bmax, Math.max(bmin, sparse[i]))
  }
  return { reconstructed: sparse, retained }
}

export const BROWSER_WAVELET = 'haar'
export const WAVELET_FAMILIES = ['haar', 'db2', 'db4', 'sym2', 'sym4'] as const
export type WaveletFamily = (typeof WAVELET_FAMILIES)[number]

export function runWaveletCompression(
  bands: BandMap,
  bandOrder: string[],
  width: number,
  height: number,
  options: { keepFraction: number; levels: number; wavelet?: string },
): { bands: BandMap; metadata: Record<string, unknown>; compressedBytesEstimate: number } {
  const out: BandMap = {}
  let retainedTotal = 0
  for (const name of bandOrder) {
    const result = compressBandWavelet(
      bands[name],
      width,
      height,
      options.keepFraction,
      options.levels,
    )
    out[name] = result.reconstructed
    retainedTotal += result.retained
  }
  return {
    bands: out,
    metadata: {
      wavelet: BROWSER_WAVELET,
      requestedWavelet: options.wavelet || BROWSER_WAVELET,
      keepFraction: options.keepFraction,
      levels: options.levels,
      preserveApproximation: true,
      backend: 'haar-multilevel-ll',
      note:
        options.wavelet && options.wavelet !== 'haar'
          ? `Browser engine uses Haar; select Cloud Run for ${options.wavelet}`
          : undefined,
    },
    compressedBytesEstimate: retainedTotal * 8,
  }
}
